File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
/*
keeping a map to store last seen index for a character
left pointer to start window
iterate right pointer over string
if character seen inside window, move right
update max length
*/

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char,int> mp;

        int left=0;
        int maxLen=0;

        for(int right=0;right<s.size();right++){
            char ch=s[right];
            if(mp.find(ch)!=mp.end()&&mp[ch]>=left){
                left=mp[ch]+1;
            }

            mp[ch]=right;
            maxLen=max(maxLen,right-left+1);
        }
        return maxLen;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        left=0 #left pointer of window
        char_set=set() #to store unique characters
        max_length=0

        #expanding window using right pointer
        for right in range(len(s)):

            while s[right] in char_set:
                char_set.remove(s[left])
                left+=1
            
            char_set.add(s[right])

            max_length=max(max_length,right-left+1)
        return max_length
        